### PR TITLE
ARROW-11504: [Rust] Added checks to List DataType.

### PR DESCRIPTION
**Pull request by jorgecarleitao**
This PR fixes a bug on which `GenericListArray` is not validating the datatype passed on to `from(ArrayData)`, causing all types of bugs, such as undefined behavior in interpreting the offset buffer.

This PR adds this validation, panicking if the DataType does not match.

This PR also fixes casting from and to Lists, which was creating an `ArrayData` out of spec.

### CODE REVIEW

File: rust/arrow/src/array/array_list.rs
Comment (sunchao) [inline rust/arrow/src/array/array_list.rs:172]: can we have a few tests to cover this? checking error message and all. 

Also I'm not sure if `assert_eq` is good here:  IMO assertion should only be used for checking internal logic that developer should follow and which are not exposed to the library users, but in this case it appears not. It's just a nit though since this is already used in multiple places before.

File: rust/arrow/src/array/array_list.rs
Comment (sunchao) [inline rust/arrow/src/array/array_list.rs:47]: maybe we won't need `prefix` anymore with the new `is_large`.

**Review by sunchao — COMMENTED** [review]
Thanks @jorgecarleitao for pinging. Left a few comments.

**Comment by jhorstmann** [comment]
Nice! I recently ran into the same issue and the assertion about the nested datatypes would have saved me some time debugging. In my testcase an assert_eq was failing, but printed both sides exactly the same.

**Comment by jorgecarleitao** [comment]
@jhorstmann , I think that that shows another, separate issue: imo what we currently show in "debug" should be shown in "Display", and "Debug" should actually show the full structure, i.e. the one created by `#[derive(Debug)]`.

File: rust/arrow/src/array/array_list.rs
Comment (nevi-me) [reply rust/arrow/src/array/array_list.rs:47]: We might still need it, we also use it for formatting in `Display`

**Review by nevi-me — COMMENTED** [review]
No additional comments, other than @sunchao's

**Review by alamb — COMMENTED** [review]
@jorgecarleitao  do you need help with this PR? I can try and take some of @sunchao 's comments if that would help

File: rust/arrow/src/array/array_list.rs
Comment (jorgecarleitao) [reply rust/arrow/src/array/array_list.rs:47]: I think that we can drop it, yes. We can merge StringOffset, BinaryOffset and OffsetTrait in a single Trait with this, but I wanted to leave it to another PR.

File: rust/arrow/src/array/array_list.rs
Comment (jorgecarleitao) [reply rust/arrow/src/array/array_list.rs:172]: I agree with you. However, that requires a larger change as we would need to move from `From` to `TryFrom`, so for now I just want to avoid `unsafe` code by panicking everytime something may go wrong.

**Comment by jorgecarleitao** [comment]
@alamb that would definitely help. If you have the time, I would really appreciate.

**Comment by alamb** [comment]
> @alamb that would definitely help. If you have the time, I would really appreciate.

@jorgecarleitao  I will put it on my queue for tomorrow

File: rust/arrow/src/array/array_list.rs
Comment (alamb) [reply rust/arrow/src/array/array_list.rs:47]: Here is one way to remove `prefix` that does not go as far as @jorgecarleitao  suggests to collapse the traits... https://github.com/apache/arrow/pull/9508/commits/8e68e05160a56f9feb64438e4dcd739a67a884e1

**Review by alamb — APPROVED** [review]
I think this PR is good to merge as is as it makes the code better than on master (less errors), though it can be further improved. 

@sunchao  when you get some time I would love your feedback / advice -- should we merge this PR as is? Would you suggest incorporating one/both of the approaches prototyped in  https://github.com/apache/arrow/pull/9508 (remove `prefix` and making a fallable version of `ArrayData` --> `GenericListArray`)

**Review by sunchao — APPROVED** [review]
LGTM2 - I think we can merge this as it is and solve the comments separately. Thanks!

BTW this needs rebase.

**Comment by nevi-me** [comment]
@alamb can we merge this when CI is green, then treat #9508 as a separate JIRA ticket? This had fallen quite behind, but I've rebased and fixed failures on it, when I should have done it against #9508 instead.

**Comment by alamb** [comment]
@nevi-me  good plan!

File: rust/arrow/src/ipc/writer.rs
Comment (alamb) [inline rust/arrow/src/ipc/writer.rs:987]: @nevi-me I don't remember seeing this in the original PR -- was this change intended ?

Comment (alamb) [reply rust/arrow/src/ipc/writer.rs:987]: NM I see https://github.com/apache/arrow/pull/9587 now

**Pull request closed without merging**
